docs/developer_guide/node_implementation.md

```diff
@@ -169,14 +169,61 @@ The timeout argument in these methods signify the duration in which the method h
 
 
 ### Logging
-
-Node classes inherit a method named `self.log` that allows users to log messages. You can set the appropriate level from: `info`, `debug`, `warning`, `error`, an `critical`. You can log from inside of the `_pre_loop_hook` method, the `_execute` method, or any other method you add to your node.
+#### From within a node
+Node classes inherit a method named `self.log` that allows users to produce messages to the `logging` dataset. This is useful for debugging and monitoring purposes. You can specify the level of the log message, by specifying the `level` argument. By default, the level is set to `info`. Other valid levels are `debug`, `warning`, `error`, and `critical`. You can log from any method that inherits from `AbstractNode`.
 
 :   
     ```python
     self.log(f"Produced {self.cur_integer}", level="info")
     ```
 
+#### From outside a node
+Since `self.log` is inherited from `AbstractNode`, you can't use it outside of a node. However, you can still log to the `logging` dataset by using the native Python `logging` module and specifying the `logging_namespace` and setting `log_to_dataset` to `true` in the node or pipeline configuration as shown below. This configuration will read all log entries from the "foo" namespace and produce them to the `logging` dataset.
+
+:   
+
+    ```yaml title="conf/pipeline.yml" linenums="7"
+    nodes:
+      example_node:
+        class: your_package.your_module.YourNode
+        log_to_dataset: true
+        logging_namespace: foo
+    ```
+
+    ```python title="your_package/your_utils.py" linenums="7"
+    import logging
+
+    logger = logging.getLogger("foo")
```

Comment:
I think it's confusing that people have to create a logger namespace in a node before they can use it anywhere. Can we do something like creating a default namespace called `kafka_logging` or `logging` and create `logger = logging.getLogger("foo")` in `aineko.__init__.py` so users can do `from aineko import logging` without having to create it first. 

This doesn't stop them from creating their own loggers.

Reply:
maybe we can use the NodeManager or one of the internal node's self.log method for this. Or implement an independent logger that doesn't require a self.log

Reply:
If we make it dependent on a `self.log` method call, this will be unable to capture logs that occur before a pipeline goes up and after a pipeline goes down. I think having an independent dataset producer object that writes to the logging topic might be better

Reply:
> I think it's confusing that people have to create a logger namespace in a node before they can use it anywhere. 

Afaik it's a common/best practice to initialise a logger with `logger = logging.getLogger(__name__) `  at the top of each module. This allows for hierarchical loggers as described in [Logger Objects](https://docs.python.org/3/library/logging.html#logger-objects).

```
2024-01-02 09:43:48.000 - your_package.your_module_1 - INFO - Hello from module 1!
2024-01-02 09:43:49.000 - your_package.your_module_2 - INFO - Hello from module 2!
2024-01-02 09:43:50.000 - your_package.your_module_1 - ERROR - Hello from module 1!
```

> Can we do something like creating a default namespace called `kafka_logging` or `logging` and create `logger = logging.getLogger("foo")` in `aineko/__init__.py` so users can do `from aineko import logging` without having to create it first.

I think it would make the feature less useful since users would never know where the log came from. They would all originate from "foo".

```
2024-01-02 09:43:48.458 - foo - INFO - Hello, World!
```

Reply:
> maybe we can use the NodeManager or one of the internal node's self.log method for this. Or implement an independent logger that doesn't require a self.log

> If we make it dependent on a `self.log` method call, this will be unable to capture logs that occur before a pipeline goes up and after a pipeline goes down. I think having an independent dataset producer object that writes to the logging topic might be better

How do you envision this independent object? 

The complexity of logging before/during a pipeline starting lies in the distributed nature of the processes and the creation of the topics. Everything that happens before a node starts isn't visible to the node, so it can't capture and produce. But maybe we can store these logs temporarily until the NodeManager is running and then produce them to the logging topic. It's not a trivial feature, so maybe we should spec that out separately.

Reply:
I'm not sure if this kind of hierarchy makes sense. Since you might have the same utils function being called in multiple nodes but you'll never know which was the node that called the util function and triggered this log.

Reply:
> > maybe we can use the NodeManager or one of the internal node's self.log method for this. Or implement an independent logger that doesn't require a self.log
> 
> > If we make it dependent on a `self.log` method call, this will be unable to capture logs that occur before a pipeline goes up and after a pipeline goes down. I think having an independent dataset producer object that writes to the logging topic might be better
> 
> How do you envision this independent object?
> 
> The complexity of logging before/during a pipeline starting lies in the distributed nature of the processes and the creation of the topics. Everything that happens before a node starts isn't visible to the node, so it can't capture and produce. But maybe we can store these logs temporarily until the NodeManager is running and then produce them to the logging topic. It's not a trivial feature, so maybe we should spec that out separately.

I would agree with this IF they turn out to be separate implementations, but I think if we want to stick to one consistent logging mechanism (I think we should) then the more general one, i.e. independent of nodes) should be the one. 

I agree it's more complex - the most straightforward but maybe inefficient method I can think of is for us to create a DatasetProducer in the __init__.py file as a global variable that the `emit` method will use to produce logs. The emit should retry and fail elegantly in cases the logging topic doesn't exist.